### PR TITLE
Use temporary self signed certificates for JSON-RPC HTTP TLS unit testing

### PR DESCRIPTION
**Pull request by usmansaleem**
## PR description
 -- Generate and use temporary self signed certificates in PKCS12 format with SAN extensions for dns and IP to be used by TLS unit/integration tests
 -- Remove pre-generated key stores and trust stores used by TLS tests

Signed-off-by: Usman Saleem <usman@usmans.info>

### CODE REVIEW

File: gradle/versions.gradle [hunk 1]
```diff
@@ -76,6 +76,7 @@ dependencyManagement {
 
     dependency 'org.awaitility:awaitility:4.0.1'
 
+    dependency 'org.bouncycastle:bcpkix-jdk15on:1.64'
```
Comment (rolfyone) [inline gradle/versions.gradle]: duplicated dependency?

Comment (usmansaleem) [reply gradle/versions.gradle]: Where is it duplicated? The existing one is `bcprov`, the new one in `bcpkix`.

Comment (usmansaleem) [reply gradle/versions.gradle]: Where is it duplicated? The existing one is `bcprov`, the new one in `bcpkix`.

File: ethereum/api/build.gradle [hunk 2]
```diff
@@ -27,6 +27,12 @@ jar {
   }
 }
 
+configurations { testArtifacts }
+task testJar (type: Jar) {
```
Comment (rain-on) [inline ethereum/api/build.gradle]: why does the test jar need to be published? Can this not be done with the testSupport?

File: ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedPfxStore.java [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+  private static final String DEFAULT_ALIAS = "test";
+  private static final BouncyCastleProvider BOUNCY_CASTLE_PROVIDER = new BouncyCastleProvider();
+  private static final boolean IS_CA = true;
+  private final Path parentPath;
+  private final String alias;
+  private final char[] password;
+  private final String distinguishedName;
+  private final List<String> sanHostNames;
+  private final List<String> sanIpAddresses;
+  private X509Certificate certificate;
```
Comment (rain-on) [inline ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedPfxStore.java]: nit: having 1/2 final and 1/2 not final variables is a smell that this is maybe 2 classes

File: ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedPfxStore.java [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+  private Path passwordFile;
+
+  private SelfSignedPfxStore(
+      final Path parentPath,
+      final String alias,
+      final char[] password,
+      final String distinguishedName,
+      final List<String> sanHostNames,
+      final List<String> sanIpAddresses) {
+    this.parentPath = parentPath;
```
Comment (rain-on) [inline ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedPfxStore.java]: nit: parentPath is good for lots of things - but given this is multiple files, maybe its containingFolder?
What happens if this points to a file?

File: ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedPfxStore.java [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+      throws IOException, KeyStoreException, NoSuchAlgorithmException, CertificateException {
+    final Path pfxPath = Files.createTempFile(parentPath, alias, ".pfx");
+    pfxPath.toFile().deleteOnExit();
+    try (final FileOutputStream outputStream = new FileOutputStream(pfxPath.toFile())) {
+      keyStore.store(outputStream, password);
+    }
+    return pfxPath;
+  }
+
+  private void createKnownClientsFile() throws IOException, CertificateEncodingException {
```
Comment (rain-on) [inline ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedPfxStore.java]: maybe this doesn't have a known-clients file, but rather, allows you to get the fingerprint of the certificate, that way the user of this class can choose what to do with the fingerprint - i.e. compare it, write it to file etc.etc.

File: ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedPfxStore.java [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+    final String commonName = IETFUtils.valueToString(commonNameRDN.getFirst().getValue());
+    final String fingerPrint = TLS.certificateHexFingerprint(certificate);
+
+    final Path tempFile = Files.createTempFile(parentPath, alias + "knownClientsFile", ".txt");
+    tempFile.toFile().deleteOnExit();
+    Files.writeString(tempFile, commonName + " " + fingerPrint);
+    this.knownClientsFile = tempFile;
+  }
+
+  private void createPasswordFile() throws IOException {
```
Comment (rain-on) [inline ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedPfxStore.java]: nit: does a certificate need a password _file_ - or just a password? Regarding abstractions, is it useful to get this key's password, and let the user of this class write it to file?

File: ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedPfxStore.java [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+    final Path tempFile = Files.createTempFile(parentPath, alias + "pass", ".txt");
```
Comment (rain-on) [inline ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedPfxStore.java]: is createTempFile better/worse/different to performing a Path.resolve?

File: ethereum/api/build.gradle [hunk 8]
```diff
@@ -56,6 +62,8 @@ dependencies {
   implementation 'org.bouncycastle:bcprov-jdk15on'
   implementation 'org.springframework.security:spring-security-crypto'
 
+  runtimeOnly 'org.bouncycastle:bcpkix-jdk15on'
```
Comment (CjHare) [inline ethereum/api/build.gradle]: Why is this needed, the `crypto` project already has bouncy castle as an `api` dependency?

File: ethereum/api/build.gradle [hunk 9]
```diff
@@ -70,6 +78,8 @@ dependencies {
   testImplementation 'org.assertj:assertj-core'
   testImplementation 'org.mockito:mockito-core'
 
+  testSupportImplementation 'org.bouncycastle:bcpkix-jdk15on'
```
Comment (CjHare) [inline ethereum/api/build.gradle]: Are you sure this addition is needed? ...as `crypto` already has the dependency as an `api`, doesn't result it already being on the path for test support?

File: ethereum/api/build.gradle (hunk 2 above)
Comment (usmansaleem) [reply ethereum/api/build.gradle]: gone. It wasn't required.

File: ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedPfxStore.java (hunk 4 above)
Comment (usmansaleem) [reply ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedPfxStore.java]: Refactored and simplified.

File: ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedPfxStore.java (hunk 3 above)
Comment (usmansaleem) [reply ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedPfxStore.java]: Refactored and simplified.

File: ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedPfxStore.java (hunk 5 above)
Comment (usmansaleem) [reply ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedPfxStore.java]: Extracted out to a different utility class. It now determines common name and fingerprint dynamically from trust store.

File: ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedPfxStore.java (hunk 6 above)
Comment (usmansaleem) [reply ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedPfxStore.java]: Refactored.

File: ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedPfxStore.java (hunk 7 above)
Comment (usmansaleem) [reply ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedPfxStore.java]: Although its out now, createTempFile will always create a unique file name. Resolve is useful where you want known file name.

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcHttpServiceTlsTest.java [hunk 10]
```diff
@@ -83,25 +86,19 @@
...
   private final JsonRpcTestHelper testHelper = new JsonRpcTestHelper();
+  private SelfSignedPfxStore selfSignedPfxStore;
+  private SelfSignedPfxStore selfSignedPfxStoreForHttpClient;
+  private SelfSignedPfxStore selfSignedPfxStoreForHttpClientNotTrustedByServer;
 
   @Before
   public void initServer() throws Exception {
+    selfSignedPfxStore = SelfSignedPfxStore.create();
+    selfSignedPfxStoreForHttpClient = SelfSignedPfxStore.create();
+    selfSignedPfxStoreForHttpClientNotTrustedByServer = SelfSignedPfxStore.create();
```
Comment (CjHare) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcHttpServiceTlsTest.java]: This seems odd, the same method of creation gives a `PfxStore`, `PfxStoreForHttpClient` and `PfxStoreForHttpClientNotTrustedByServer`. 

If there are three different stores, then perhaps separate helper methods to create each of them and their corresponding state (e.g. that makes one trusted and another not trusted), might be an idea?

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcHttpServiceTlsMisconfigurationTest.java [hunk 11]
```diff
@@ -121,6 +122,15 @@ public void beforeEach() {
                     Collections.emptyMap()));
   }
 
+  private Path createKnownClientsFile() throws Exception {
```
Comment (CjHare) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcHttpServiceTlsMisconfigurationTest.java]: Is this method a duplicate also added in `JsonRpcHttpServiceTlsTest.java`, util method for reuse?

File: ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/TrustStoreUtil.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+import org.apache.tuweni.net.tls.TLS;
+import org.bouncycastle.asn1.x500.RDN;
+import org.bouncycastle.asn1.x500.X500Name;
+import org.bouncycastle.asn1.x500.style.BCStyle;
+import org.bouncycastle.asn1.x500.style.IETFUtils;
+import org.bouncycastle.cert.X509CertificateHolder;
+
+public class TrustStoreUtil {
+  public static String commonNameAndFingerPrint(
```
Comment (CjHare) [inline ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/TrustStoreUtil.java]: keep consistent naming, either prefix both methods with `get` or neither.

**Comment by usmansaleem** [comment]
> What was the motivation for locating this self signing code under ethereum/api rather then say crypto?

@CjHare The test code (and the unit under testing) which required self signed certificates was in ethereum/api. I don't mind moving it under crypto/testSupport (I think it make more sense there)

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcHttpServiceTlsTest.java (hunk 10 above)
Comment (usmansaleem) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcHttpServiceTlsTest.java]: The utility class generate both keystore and truststore containing self-sligned certificate. The keystore is used by server/json-rpc and truststore is used by okhttp/client vice-versa is also applicable for testing client-auth. The keystore used by okhttp client, its truststore is used by server. Hence there is a need of having three different instances of self-signed certs. I can wrap them in a method for the sake of reaability, but all it will do is return instances which will again store in three different variables.

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcHttpServiceTlsMisconfigurationTest.java (hunk 11 above)
Comment (usmansaleem) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcHttpServiceTlsMisconfigurationTest.java]: let me check. Probably a candidate to move in common testSupport or utility class.

File: ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/TrustStoreUtil.java (hunk 12 above)
Comment (usmansaleem) [reply ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/TrustStoreUtil.java]: agreed.

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcHttpServiceTlsMisconfigurationTest.java (hunk 11 above)
Comment (usmansaleem) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcHttpServiceTlsMisconfigurationTest.java]: This has been moved to a separate utility class.

File: ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/TrustStoreUtil.java (hunk 12 above)
Comment (usmansaleem) [reply ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/TrustStoreUtil.java]: The class name has been refactored to clear the intent. Also, methods has been changed accordingly as well.

File: ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/KnownClientFileUtil.java [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+public class KnownClientFileUtil {
+
+  public static Path createKnownClientsFile(
+      final SelfSignedP12Certificate selfSignedP12Certificate) {
+    try {
+      final Certificate certificate = selfSignedP12Certificate.getCertificate();
+      final String fingerprint = TLS.certificateHexFingerprint(certificate);
+      final String commonName = extractCommonName(certificate);
+      final String knownClientsLine = String.format("%s %s", commonName, fingerprint);
+      final Path temporaryKnownClientsFile = Files.createTempFile("testKnownClients", ".txt");
```
Comment (rain-on) [inline ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/KnownClientFileUtil.java]: nit: I'd remove "test" - you never know if someone wants to make this a general capability in an app/library.
Having said that - having this create a tempFile, when you don't _know_ the use of this function is kinda dangerous.
But am wondering if these two functions should effeictvely exist on the SelfSignedP12Certificate - such taht someone can say:
cert.getFingerPrint();
cert.getCommonName();
writeFile(fingerprint, commonName)

File: ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedP12Certificate.java [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+import org.bouncycastle.operator.OperatorCreationException;
+import org.bouncycastle.operator.jcajce.JcaContentSignerBuilder;
+
+/**
+ * Utility class to generate temporary self-signed certificates in PKCS12 format for testing
+ * purposes using BouncyCastle APIs. The generated certificate supports SAN extension for multiple
+ * DNS and IP addresses
+ *
+ * <p>Note: DO NOT USE IN PRODUCTION. The generated stores and files are marked to be deleted on JVM
+ * exit.
```
Comment (rain-on) [inline ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedP12Certificate.java]: does this need a closing </p>? I know nothing of html

File: ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedP12Certificate.java [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+
+  public Path getTrustStoreFile() {
+    return trustStore;
+  }
+
+  public char[] getPassword() {
+    return password;
+  }
+
+  @SuppressWarnings("DoNotCreateSecureRandomDirectly")
```
Comment (rain-on) [inline ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedP12Certificate.java]: why are you suppressing - can this not use the secureRandom factory?

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcHttpServiceTlsMisconfigurationTest.java [hunk 16]
```diff
@@ -76,10 +76,9 @@
   private static final String CLIENT_VERSION = "TestClientVersion/0.1.0";
   private static final BigInteger CHAIN_ID = BigInteger.valueOf(123);
   private static final Collection<RpcApi> JSON_RPC_APIS = List.of(ETH, NET, WEB3);
-  private static final String KEYSTORE_RESOURCE = "JsonRpcHttpService/rpc_keystore.pfx";
-  private static final String KNOWN_CLIENTS_RESOURCE = "JsonRpcHttpService/rpc_known_clients.txt";
   private static final NatService natService = new NatService(Optional.empty());
-
+  private final SelfSignedP12Certificate besuCertificate = SelfSignedP12Certificate.create();
```
Comment (rain-on) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcHttpServiceTlsMisconfigurationTest.java]: nit: is besuCertificate correct here, or is this the certificate presented by the client?

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/TlsOkHttpClientBuilder.java [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.TrustManagerFactory;
+import javax.net.ssl.X509TrustManager;
+
+import okhttp3.OkHttpClient;
+
+public final class TlsOkHttpClientBuilder {
+  private SelfSignedP12Certificate besuCertificate;
```
Comment (rain-on) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/TlsOkHttpClientBuilder.java]: whats the difference between the besuCertificate and the okHttpCertificate?

File: ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedP12Certificate.java (hunk 15 above)
Comment (usmansaleem) [reply ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedP12Certificate.java]: Since we are in tests and creating not-so-secure certificates, we are ok by suppressing this warning.

File: ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedP12Certificate.java (hunk 14 above)
Comment (usmansaleem) [reply ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedP12Certificate.java]: No, even if we close it, spotlessJavaApply removed the closing p.

File: ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedP12Certificate.java (hunk 15 above)
Comment (ajsutton) [reply ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedP12Certificate.java]: Please don't suppress this warning, just use `SecureRandomFactory`.

Comment (usmansaleem) [reply ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/SelfSignedP12Certificate.java]: 👍

File: ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/KnownClientFileUtil.java (hunk 13 above)
Comment (usmansaleem) [reply ethereum/api/src/test-support/java/org/hyperledger/besu/ethereum/api/tls/KnownClientFileUtil.java]: I've refactored it.
